In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
df = pd.read_csv('heights.csv')
df.head()

name  height
0  mohan     5.9
1  maria     5.2
2  sakib     5.1
3    tao     5.5
4  virat     4.9

In [4]:
df['height']

0      5.9
1      5.2
2      5.1
3      5.5
4      4.9
5      5.4
6      6.2
7      6.5
8      7.1
9     14.5
10     6.1
11     5.6
12     1.2
13     5.5
Name: height, dtype: float64

In [5]:
max_threshold = df['height'].quantile(0.95)
max_threshold

9.689999999999998

In [6]:
df[df['height']>max_threshold]

name  height
9  imran    14.5

In [7]:
min_threshold = df['height'].quantile(0.05)
min_threshold

3.6050000000000004

In [8]:
df[df['height']<min_threshold]

name  height
12  yoseph     1.2

In [11]:
df=df[(df['height']<max_threshold) & (df['height']>min_threshold)]

In [12]:
df

name  height
0     mohan     5.9
1     maria     5.2
2     sakib     5.1
3       tao     5.5
4     virat     4.9
5    khusbu     5.4
6    dmitry     6.2
7    selena     6.5
8      john     7.1
10     jose     6.1
11  deepika     5.6
13    binod     5.5

In [13]:
# Feature engineering on Bangalore House prices data 

In [14]:
df_bhp = pd.read_csv('bhp.csv')
df_bhp

location       size  total_sqft  bath   price  bhk  \
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2   
...                         ...        ...         ...   ...     ...  ...   
13195                Whitefield  5 Bedroom      3453.0   4.0  231.00    5   
13196                     other      4 BHK      3600.0   5.0  400.00    4   
13197     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2   
13198           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4   
13199              Doddathoguru      1 BHK       550.0   1.0   17.00    1   

       price_per_sqft  
0                3699  
1                4615  
2                4305  
3                6245  
4                4250  
...               ...  
13195            6689  
13196           11111  
13197            5258  
13198           10407  
13199            3090  

[13200 rows x 7 columns]

In [15]:
df_bhp.shape

(13200, 7)

In [17]:
df_bhp.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  13200.000000  13200.000000  13200.000000  13200.000000    1.320000e+04
mean    1555.302783      2.691136    112.276178      2.800833    7.920337e+03
std     1237.323445      1.338915    149.175995      1.292843    1.067272e+05
min        1.000000      1.000000      8.000000      1.000000    2.670000e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.267000e+03
50%     1275.000000      2.000000     71.850000      3.000000    5.438000e+03
75%     1672.000000      3.000000    120.000000      3.000000    7.317000e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [19]:
min_threshold, max_threshold = df_bhp.price_per_sqft.quantile([0.001, 0.999])
min_threshold, max_threshold

(1366.184, 50959.36200000098)

So, based upon the above min & max threshold values we will consider that houses having 'price_per_sqft' less than 1366.184 and greater than 50959 as outliers

In [20]:
df_bhp[df_bhp.price_per_sqft<min_threshold]

location       size  total_sqft  bath  price  bhk  price_per_sqft
665    Yelahanka      3 BHK     35000.0   3.0  130.0    3             371
798        other  4 Bedroom     10961.0   4.0   80.0    4             729
1867       other  3 Bedroom     52272.0   2.0  140.0    3             267
2392       other  4 Bedroom      2000.0   3.0   25.0    4            1250
3934       other      1 BHK      1500.0   1.0   19.5    1            1300
5343       other      9 BHK     42000.0   8.0  175.0    9             416
5417      Ulsoor      4 BHK     36000.0   4.0  450.0    4            1250
5597    JP Nagar      2 BHK      1100.0   1.0   15.0    2            1363
7166   Yelahanka  1 Bedroom     26136.0   1.0  150.0    1             573
7862    JP Nagar      3 BHK     20000.0   3.0  175.0    3             875
8300     Kengeri      1 BHK      1200.0   1.0   14.0    1            1166
9144       other  4 Bedroom     10961.0   4.0   80.0    4             729
11635      Begur      3 BHK      2400.0   3.0   12.0    3             500
12355      other      4 BHK     16335.0   4.0  149.0    4             912

In [21]:
df_bhp[df_bhp.price_per_sqft>max_threshold]

location       size  total_sqft  bath   price  bhk  \
345                other  3 Bedroom        11.0   3.0    74.0    3   
1005               other      1 BHK        15.0   1.0    30.0    1   
1106               other  5 Bedroom        24.0   2.0   150.0    5   
4044      Sarjapur  Road  4 Bedroom         1.0   4.0   120.0    4   
4924               other      7 BHK         5.0   7.0   115.0    7   
5911         Mysore Road  1 Bedroom        45.0   1.0    23.0    1   
6356        Bommenahalli  4 Bedroom      2940.0   3.0  2250.0    4   
7012               other      1 BHK       650.0   1.0   500.0    1   
7575               other      1 BHK       425.0   1.0   750.0    1   
7799               other      4 BHK      2000.0   3.0  1063.0    4   
8307   Bannerghatta Road      5 BHK      2500.0   4.0  1400.0    5   
9436        Indira Nagar  4 Bedroom      2400.0   5.0  1250.0    4   
11447         Whitefield  4 Bedroom        60.0   4.0   218.0    4   
12328              other  4 Bedroom      4350.0   8.0  2600.0    4   

       price_per_sqft  
345            672727  
1005           200000  
1106           625000  
4044         12000000  
4924          2300000  
5911            51111  
6356            76530  
7012            76923  
7575           176470  
7799            53150  
8307            56000  
9436            52083  
11447          363333  
12328           59770

In [22]:
df_bhp2 = df_bhp[(df_bhp.price_per_sqft > min_threshold) & (df_bhp.price_per_sqft < max_threshold)]
df_bhp2

location       size  total_sqft  bath   price  bhk  \
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2   
...                         ...        ...         ...   ...     ...  ...   
13195                Whitefield  5 Bedroom      3453.0   4.0  231.00    5   
13196                     other      4 BHK      3600.0   5.0  400.00    4   
13197     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2   
13198           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4   
13199              Doddathoguru      1 BHK       550.0   1.0   17.00    1   

       price_per_sqft  
0                3699  
1                4615  
2                4305  
3                6245  
4                4250  
...               ...  
13195            6689  
13196           11111  
13197            5258  
13198           10407  
13199            3090  

[13172 rows x 7 columns]

In [23]:
df_bhp2.shape

(13172, 7)

In [24]:
df_bhp.shape

(13200, 7)

In [25]:
df_bhp2.sample(10)  # used to get random sample of 10

location       size  total_sqft  bath  price  bhk  \
11234                   other      2 BHK      1000.0   2.0   45.0    2   
10201               Yelahanka      3 BHK      1890.0   4.0   85.0    3   
10764                   other      3 BHK      3000.0   3.0  265.0    3   
9170                    other  3 Bedroom      1700.0   3.0  250.0    3   
10475              Whitefield  4 Bedroom      2500.0   4.0  100.0    4   
11128          Sarjapur  Road      2 BHK      1034.0   2.0   38.0    2   
7954           Sarjapur  Road      3 BHK      1550.0   2.0   89.0    3   
10209           Doddaballapur  3 Bedroom      2440.0   3.0  142.0    3   
6911   Kengeri Satellite Town      2 BHK       920.0   2.0   39.0    2   
5896           Dodda Nekkundi      2 BHK      1148.0   2.0   56.0    2   

       price_per_sqft  
11234            4500  
10201            4497  
10764            8833  
9170            14705  
10475            4000  
11128            3675  
7954             5741  
10209            5819  
6911             4239  
5896             4878

In [26]:
# Exercise

In [29]:
df_airbnb = pd.read_csv('AB_NYC_2019.csv')

In [30]:
df_airbnb

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

In [32]:
df_airbnb.shape

(48895, 16)

In [34]:
df_airbnb.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

In [36]:
df_airbnb.isna().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [37]:
# dropping the irrelevant columns
df_airbnb.drop(['id','name','host_id','host_name','last_review'],axis=1,inplace=True)

In [38]:
df_airbnb

neighbourhood_group       neighbourhood  latitude  longitude  \
0                Brooklyn          Kensington  40.64749  -73.97237   
1               Manhattan             Midtown  40.75362  -73.98377   
2               Manhattan              Harlem  40.80902  -73.94190   
3                Brooklyn        Clinton Hill  40.68514  -73.95976   
4               Manhattan         East Harlem  40.79851  -73.94399   
...                   ...                 ...       ...        ...   
48890            Brooklyn  Bedford-Stuyvesant  40.67853  -73.94995   
48891            Brooklyn            Bushwick  40.70184  -73.93317   
48892           Manhattan              Harlem  40.81475  -73.94867   
48893           Manhattan      Hell's Kitchen  40.75751  -73.99112   
48894           Manhattan      Hell's Kitchen  40.76404  -73.98933   

             room_type  price  minimum_nights  number_of_reviews  \
0         Private room    149               1                  9   
1      Entire home/apt    225               1                 45   
2         Private room    150               3                  0   
3      Entire home/apt     89               1                270   
4      Entire home/apt     80              10                  9   
...                ...    ...             ...                ...   
48890     Private room     70               2                  0   
48891     Private room     40               4                  0   
48892  Entire home/apt    115              10                  0   
48893      Shared room     55               1                  0   
48894     Private room     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                    NaN                               1               365  
3                   4.64                               1               194  
4                   0.10                               1                 0  
...                  ...                             ...               ...  
48890                NaN                               2                 9  
48891                NaN                               2                36  
48892                NaN                               1                27  
48893                NaN                               6                 2  
48894                NaN                               1                23  

[48895 rows x 11 columns]

In [39]:
df_airbnb.isna().sum()

neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [41]:
df_airbnb.isnull().sum()

neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [43]:
df_airbnb.reviews_per_month.mean()

1.3732214298586884

In [45]:
df_airbnb.reviews_per_month.fillna(df_airbnb.reviews_per_month.mean(), inplace = True)

In [46]:
df_airbnb.isna().sum()

neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [56]:
min_threshold, max_threshold = df_airbnb['price'].quantile([0.01,0.99])
min_threshold, max_threshold

(30.0, 799.0)

In [57]:
df_airbnb[df_airbnb.price<min_threshold]

neighbourhood_group       neighbourhood  latitude  longitude  \
957         Staten Island        Graniteville  40.62109  -74.16534   
2675             Brooklyn            Flatbush  40.63899  -73.95177   
2860            Manhattan        East Village  40.73051  -73.98140   
3020               Queens           Woodhaven  40.68968  -73.85219   
3918             Brooklyn  Bedford-Stuyvesant  40.68642  -73.93440   
...                   ...                 ...       ...        ...   
48486              Queens        Forest Hills  40.73657  -73.85088   
48647              Queens       Richmond Hill  40.68639  -73.81847   
48832            Brooklyn            Flatbush  40.64922  -73.96078   
48867            Brooklyn           Gravesend  40.59080  -73.97116   
48868            Brooklyn           Gravesend  40.59118  -73.97119   

             room_type  price  minimum_nights  number_of_reviews  \
957       Private room     20               3                 80   
2675       Shared room     29               2                  5   
2860   Entire home/apt     10              30                  0   
3020      Private room     29               2                386   
3918       Shared room     25               5                 76   
...                ...    ...             ...                ...   
48486  Entire home/apt     16               9                  1   
48647     Private room     28               2                  0   
48832      Shared room     20               1                  0   
48867      Shared room     25               1                  0   
48868      Shared room     25               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
957             0.920000                               1               226  
2675            0.070000                               1               321  
2860            1.373221                               4               137  
3020            5.530000                               3                50  
3918            1.220000                               3               258  
...                  ...                             ...               ...  
48486           1.000000                               1               322  
48647           1.373221                               1                24  
48832           1.373221                               3               363  
48867           1.373221                               6               338  
48868           1.373221                               6               365  

[404 rows x 11 columns]

In [59]:
df_airbnb[df_airbnb.price>max_threshold]

neighbourhood_group       neighbourhood  latitude  longitude  \
85               Brooklyn    Brooklyn Heights  40.69723  -73.99268   
496             Manhattan     Upper West Side  40.77350  -73.98697   
662             Manhattan   Greenwich Village  40.73129  -73.99944   
762             Manhattan        West Village  40.73301  -74.00268   
780              Brooklyn  Bedford-Stuyvesant  40.68492  -73.95489   
...                   ...                 ...       ...        ...   
48304           Manhattan      Hell's Kitchen  40.75533  -73.99866   
48305           Manhattan     Upper East Side  40.78132  -73.95262   
48523            Brooklyn  Bedford-Stuyvesant  40.68767  -73.95805   
48535            Brooklyn  Bedford-Stuyvesant  40.68245  -73.93417   
48839            Brooklyn      Sheepshead Bay  40.59866  -73.95661   

             room_type  price  minimum_nights  number_of_reviews  \
85     Entire home/apt    800               1                 25   
496    Entire home/apt   2000              30                 30   
662    Entire home/apt    850               3                107   
762    Entire home/apt   1300               5                 28   
780    Entire home/apt    800               4                122   
...                ...    ...             ...                ...   
48304  Entire home/apt   2999              30                  0   
48305  Entire home/apt   1999              30                  0   
48523  Entire home/apt   1369               1                  0   
48535  Entire home/apt   1749               1                  0   
48839     Private room    800               1                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
85              0.240000                               1                 7  
496             0.330000                              11                 0  
662             1.150000                               1               249  
762             0.310000                               4               297  
780             1.370000                               1               257  
...                  ...                             ...               ...  
48304           1.373221                               1               222  
48305           1.373221                               1               270  
48523           1.373221                               1               349  
48535           1.373221                               1               303  
48839           1.373221                               1                23  

[474 rows x 11 columns]

In [60]:
df_airbnb = df_airbnb[(df_airbnb.price>min_threshold) & (df_airbnb.price<max_threshold)]

In [61]:
df_airbnb

neighbourhood_group       neighbourhood  latitude  longitude  \
0                Brooklyn          Kensington  40.64749  -73.97237   
1               Manhattan             Midtown  40.75362  -73.98377   
2               Manhattan              Harlem  40.80902  -73.94190   
3                Brooklyn        Clinton Hill  40.68514  -73.95976   
4               Manhattan         East Harlem  40.79851  -73.94399   
...                   ...                 ...       ...        ...   
48890            Brooklyn  Bedford-Stuyvesant  40.67853  -73.94995   
48891            Brooklyn            Bushwick  40.70184  -73.93317   
48892           Manhattan              Harlem  40.81475  -73.94867   
48893           Manhattan      Hell's Kitchen  40.75751  -73.99112   
48894           Manhattan      Hell's Kitchen  40.76404  -73.98933   

             room_type  price  minimum_nights  number_of_reviews  \
0         Private room    149               1                  9   
1      Entire home/apt    225               1                 45   
2         Private room    150               3                  0   
3      Entire home/apt     89               1                270   
4      Entire home/apt     80              10                  9   
...                ...    ...             ...                ...   
48890     Private room     70               2                  0   
48891     Private room     40               4                  0   
48892  Entire home/apt    115              10                  0   
48893      Shared room     55               1                  0   
48894     Private room     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0               0.210000                               6               365  
1               0.380000                               2               355  
2               1.373221                               1               365  
3               4.640000                               1               194  
4               0.100000                               1                 0  
...                  ...                             ...               ...  
48890           1.373221                               2                 9  
48891           1.373221                               2                36  
48892           1.373221                               1                27  
48893           1.373221                               6                 2  
48894           1.373221                               1                23  

[47744 rows x 11 columns]

In [62]:
df_airbnb.shape

(47744, 11)